lib/rpmsg/rpmsg_virtio.c

```diff
@@ -260,92 +261,119 @@ static int _rpmsg_virtio_get_buffer_size(struct rpmsg_virtio_device *rvdev)
 	int tick_count;
-	uint32_t buff_len;
 	int status;
-	struct metal_io_region *io;
 
 	/* Get the associated remote device for channel. */
 	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
 
-	status = rpmsg_virtio_get_status(rvdev);
 	/* Validate device state */
-	if (!(status & VIRTIO_CONFIG_STATUS_DRIVER_OK)) {
-		return RPMSG_ERR_DEV_STATE;
-	}
+	status = rpmsg_virtio_get_status(rvdev);
+	if (!(status & VIRTIO_CONFIG_STATUS_DRIVER_OK))
+		return NULL;
 
 	if (wait)
 		tick_count = RPMSG_TICK_COUNT / RPMSG_TICKS_PER_INTERVAL;
 	else
 		tick_count = 0;
 
 	while (1) {
-		int avail_size;
-
 		/* Lock the device to enable exclusive access to virtqueues */
 		metal_mutex_acquire(&rdev->lock);
-		avail_size = _rpmsg_virtio_get_buffer_size(rvdev);
-		if (avail_size && len > avail_size) {
-			metal_mutex_release(&rdev->lock);
-			return RPMSG_ERR_BUFF_SIZE;
-		}
-
-		buffer = rpmsg_virtio_get_tx_buffer(rvdev, &buff_len, &idx);
+		rp_hdr = rpmsg_virtio_get_tx_buffer(rvdev, len, &idx);
 		metal_mutex_release(&rdev->lock);
-		if (buffer || !tick_count)
+		if (rp_hdr || !tick_count)
 			break;
-		if (avail_size != 0)
-			return RPMSG_ERR_BUFF_SIZE;
 		metal_sleep_usec(RPMSG_TICKS_PER_INTERVAL);
 		tick_count--;
 	}
-	if (!buffer)
-		return RPMSG_ERR_NO_BUFF;
+
+	if (!rp_hdr)
+		return NULL;
+
+	/* Store the index into the reserved field to be used when sending */
+	rp_hdr->reserved = idx;
+
+	/* Actual data buffer size is vring buffer size minus header length */
+	*len -= sizeof(struct rpmsg_hdr);
+	return RPMSG_LOCATE_DATA(rp_hdr);
+}
+
+static int rpmsg_virtio_send_offchannel_nocopy(struct rpmsg_device *rdev,
+					       uint32_t src, uint32_t dst,
+					       const void *data, int len)
+{
+	struct rpmsg_virtio_device *rvdev;
+	struct metal_io_region *io;
+	struct rpmsg_hdr rp_hdr;
+	struct rpmsg_hdr *hdr;
+	uint32_t buff_len;
+	uint16_t idx;
+	int status;
+
+	/* Get the associated remote device for channel. */
+	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
+
+	hdr = RPMSG_LOCATE_HDR(data);
+	/* The reserved field contains buffer index */
+	idx = hdr->reserved;
 
 	/* Initialize RPMSG header. */
 	rp_hdr.dst = dst;
 	rp_hdr.src = src;
 	rp_hdr.len = len;
 	rp_hdr.reserved = 0;
+	rp_hdr.flags = 0;
 
 	/* Copy data to rpmsg buffer. */
 	io = rvdev->shbuf_io;
-	status = metal_io_block_write(io, metal_io_virt_to_offset(io, buffer),
+	status = metal_io_block_write(io, metal_io_virt_to_offset(io, hdr),
 				      &rp_hdr, sizeof(rp_hdr));
 	RPMSG_ASSERT(status == sizeof(rp_hdr), "failed to write header\r\n");
 
-	status = metal_io_block_write(io,
-				      metal_io_virt_to_offset(io,
-						RPMSG_LOCATE_DATA(buffer)),
-				      data, len);
-	RPMSG_ASSERT(status == len, "failed to write buffer\r\n");
 	metal_mutex_acquire(&rdev->lock);
 
 	/* Enqueue buffer on virtqueue. */
-	status = rpmsg_virtio_enqueue_buffer(rvdev, buffer, buff_len, idx);
+	buff_len = virtqueue_get_buffer_length(rvdev->svq, idx);
```

Comment:
Always reading the length of the buffer from the buffer descriptor is not correct. When a buffer is allocated for the first time by the function rpmsg_virtio_shm_pool_get_buffer, the buffer descriptor does not exist.

This is the cause of the segment fault.